x/ecocredit/client/testsuite/tx.go

```diff
@@ -1426,6 +1427,126 @@ func (s *IntegrationTestSuite) TestTxBuyDirect() {
+	s.Require().NoError(err)
+	buyerAddr := buyerAcc.GetAddress()
+
+	validAskDenom := sdk.DefaultBondDenom
+	askCoin := sdk.NewInt64Coin(validAskDenom, 10)
+
+	s.fundAccount(clientCtx, val0.Address, buyerAddr, sdk.Coins{sdk.NewInt64Coin(validAskDenom, 500)})
+
+	expiration, err := types.ParseDate("expiration", "3020-04-15")
+	s.Require().NoError(err)
+	_, _, batchDenom := s.createClassProjectBatch(clientCtx, valAddrStr)
+	orderIds, err := s.createSellOrder(clientCtx, &marketplace.MsgSell{
+		Owner: valAddrStr,
+		Orders: []*marketplace.MsgSell_Order{
+			{batchDenom, "10", &askCoin, true, &expiration},
+			{batchDenom, "10", &askCoin, false, &expiration},
+		},
+	})
+
+	buyQty := math.NewDecFromInt64(20)
+	buyOrders := []*marketplace.MsgBuyDirect_Order{
+		{SellOrderId: orderIds[0], Quantity: "10", BidPrice: &askCoin, DisableAutoRetire: true},
+		{SellOrderId: orderIds[1], Quantity: "10", BidPrice: &askCoin, RetirementJurisdiction: "US-OR"},
+	}
+	ordersBz, err := json.Marshal(buyOrders)
+	s.Require().NoError(err)
+	jsonFile := testutil.WriteToNewTempFile(s.T(), string(ordersBz))
+
+	makeArgs := func(fileName, from string) []string {
+		args := []string{fileName, makeFlagFrom(from)}
+		return append(args, s.commonTxFlags()...)
+	}
+
+	testCases := []struct {
+		name   string
+		args   []string
+		errMsg string
+	}{
+		{
+			name:   "too many args",
+			args:   []string{"foo", "bar"},
+			errMsg: "accepts 1 arg(s), received 2",
+		},
+		{
+			name:   "invalid: file does not exist",
+			args:   []string{"monkey.jpeg"},
+			errMsg: "no such file or directory",
+		},
+		{
+			name: "valid order",
+			args: makeArgs(jsonFile.Name(), buyerAddr.String()),
+		},
+	}
+	for _, tc := range testCases {
+		s.Run(tc.name, func() {
+			if len(tc.errMsg) != 0 {
+				_, err := cli.ExecTestCLICmd(clientCtx, cmd, tc.args)
+				s.Require().ErrorContains(err, tc.errMsg)
+			} else {
+				sellerAccBefore := s.getAccountInfo(clientCtx, val0.Address, askCoin.Denom, batchDenom)
+				buyerAccBefore := s.getAccountInfo(clientCtx, buyerAddr, askCoin.Denom, batchDenom)
+
+				costDec, err := math.NewDecFromString(askCoin.Amount.String())
+				s.Require().NoError(err)
+				totalCostDec, err := buyQty.Mul(costDec)
+				s.Require().NoError(err)
+				totalCostInt := totalCostDec.SdkIntTrim()
+				totalCostCoin := sdk.NewCoin(askCoin.Denom, totalCostInt)
+
+				out, err := cli.ExecTestCLICmd(clientCtx, cmd, tc.args)
+				s.Require().NoError(err)
+
+				sellerAccAfter := s.getAccountInfo(clientCtx, val0.Address, askCoin.Denom, batchDenom)
+				buyerAccAfter := s.getAccountInfo(clientCtx, buyerAddr, askCoin.Denom, batchDenom)
+
+				var res sdk.TxResponse
+				s.Require().NoError(clientCtx.Codec.UnmarshalJSON(out.Bytes(), &res))
+				s.Require().Equal(uint32(0), res.Code)
+
+				amtSoldEach := math.NewDecFromInt64(10)
+				s.assertMarketBalanceBatchUpdated(sellerAccBefore, sellerAccAfter, buyerAccBefore, buyerAccAfter, buyQty, totalCostCoin, amtSoldEach, amtSoldEach)
```

Comment:
Also, not sure if we can reduce the params here. `assertMarketBalanceBatchUpdated` is only used within this scenario so it might not be necessary to abstract.

Reply:
switched it to just take the marketplace orders slice. i can lump it all in one test but prefer to put it in its own function for readability purposes 🤷🏻

Reply:
Separate is ok.